### PR TITLE
fix(sources/wsl): no error with empty .cloud-init dir (SC-1862)

### PR DESCRIPTION
**Pull request by aciba90**
## Proposed Commit Message

```
fix(sources/wsl): no error with empty .cloud-init dir (SC-1862)

Do not treat the emptiness of .cloud-init/ as an error in the logs
if agent.yaml is present.

Fixes GH-5632
```

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Comment by aciba90** [comment]
I am proposing this to fix GH-5632. Could you, @CarlosNihelton, do a review, please?

**Comment by CarlosNihelton** [comment]
Hi @aciba90 ! Thanks for tackling this. I think this proposal is not the best approach because we'd be deciding how severe the error condition is inside the `find_user_data_file`. Consider the original snippet depicted below. Since `ds-identify` would disable this datasource if there is no `user_data` nor `agent_data`, then I understand that it would be an error to have this datasource running, thus raising the exception and logging with `LOG.error` is the correct behavior IMHO. On the other hand, having **only** `agent_data` should not be an error, as there is work for cloud-init to do. Under such condition we should `LOG.info`or maybe `LOG.debug`. Wether `find_user_data_file` should return None instead of raising an IOError is a matter of taste to some extent, but we certainly cannot decide there whether that's hard error or not.

```python
 # # Load Ubuntu Pro configs only on Ubuntu distros
        if self.distro.name == "ubuntu":
            agent_data, user_data = load_ubuntu_pro_data(user_home)
        # Load regular user configs
        try:
            if user_data is None and seed_dir is not None:
                user_data = ConfigData(self.find_user_data_file(seed_dir))

        except (ValueError, IOError) as err:
            LOG.error(
                "Unable to load any user-data file in %s: %s",
                seed_dir,
                str(err),
            )
```

By knowing that, I'd have implemented the fix somewhat like this:

```python
 # # Load Ubuntu Pro configs only on Ubuntu distros
        if self.distro.name == "ubuntu":
            agent_data, user_data = load_ubuntu_pro_data(user_home)
        # Load regular user configs
        try:
            if user_data is None and seed_dir is not None:
                user_data = ConfigData(self.find_user_data_file(seed_dir))

        except (ValueError, IOError) as err:
           log = LOG.info
           if agent_data is None:
               log = LOG.error
               
            log(
                "Unable to load any user-data file in %s: %s",
                seed_dir,
                str(err),
            )
```

**Review by blackboxsw — CHANGES_REQUESTED** [review]
Thanks @aciba90!  This approach seeems reasonable, but I think we may need a doc update because the documentation on the WSL datasource states per user-data files that [" If none are found, cloud-init remains disabled if no other configurations from previous steps were found."](https://cloudinit.readthedocs.io/en/latest/reference/datasources/wsl.html#user-data-configuration)..  We also need to ensure that this policy of just agent.yaml is "enough" to indicate cloud-init should detect WSL instances.

If that original documented behavior were to be retained, the WSL datasource should have actually raised
InvalidMetaDataException after this error log to avoid detection of WSL datasource altogether. 

My understanding was that even if WSL agent's agent.yaml exists, if there is no landscape user-data at .ubuntupro/.cloud-init/<instancename>.user-data then the WSL datasource MUST find minimally at least one applicable user-data file in <PROFILE_DIR>/.cloud-init/default.user-data  even if it was empty `#cloud-config\n{}`.

 Maybe this behavior/expectation/policy needs to change and be aligned with your approach in this PR. But, if so, we also need a docs change to indicate that fact that user-data is optional if minimally we have agent.yaml.    Note: this is also how ds-identify behaves (if it happens to see either .ubuntupro/.cloud-init/<INSTANCE_NAME>.user-data or agent,yaml it claims DS_FOUND, which doesn't exactly align with the python code which emits that error log prior to this PR.

**Comment by CarlosNihelton** [comment]
> Maybe this behavior/expectation/policy needs to change and be aligned with your approach in this PR. But, if so, we also need a docs change to indicate that fact that user-data is optional if minimally we have agent.yaml. Note: this is also how ds-identify behaves (if it happens to see either .ubuntupro/.cloud-init/<INSTANCE_NAME>.user-data or agent,yaml it claims DS_FOUND, which doesn't exactly align with the python code which emits that error log prior to this PR.

The `agent_data` may contain enough input for cloud-init to work on, thus that's why ds-identify claims DS_FOUND if it's found even if no other user-data file is found, thus the datasource needs to be aligned with that. We'll need to clarify that in the documentation as you've said @blackboxsw. I don't see it saying the opposite, but it's not clear enough anyway.

**Comment by blackboxsw** [comment]
> The `agent_data` may contain enough input for cloud-init to work on, thus that's why ds-identify claims DS_FOUND if it's found even if no other user-data file is found, thus the datasource needs to be aligned with that. We'll need to clarify that in the documentation as you've said @blackboxsw. I don't see it saying the opposite, but it's not clear enough anyway.


Thanks for clarifying here @CarlosNihelton. I had thought, the statement `If none are found, cloud-init remains disabled if no other configurations from previous steps were found.` was a statement applying only to the local .cloud-init/* user-data files.

Do we think this doc diff is more clear?
```diff
--- a/doc/rtd/reference/datasources/wsl.rst
+++ b/doc/rtd/reference/datasources/wsl.rst
@@ -51,7 +51,7 @@ User data can be supplied in any
 :ref:`format supported by cloud-init<user_data_formats>`, such as YAML
 cloud-config files or shell scripts. At runtime, the WSL datasource looks for
 user data in the following locations inside the Windows host filesystem, in the
-order specified below.
+order specified below. The WSL datasource will not be enabled if cloud-init does not discover any of these applicable config files below.
 
 First, configurations from Ubuntu Pro/Landscape are checked for in the
 following paths:
@@ -71,8 +71,8 @@ following paths:
    used instead of the one provided by the ``agent.yaml``, which is treated as
    a default.
 
-Then, if a file from (1) is not found, a user-provided configuration will be
-looked for instead in the following order:
+Then, if a file from (1) is not found, optional user-provided configuration
+will be looked for in the following order:
 
 1. ``%USERPROFILE%\.cloud-init\<InstanceName>.user-data`` holds user data for a
    specific instance configuration. The datasource resolves the name attributed
```

**Comment by CarlosNihelton** [comment]
@blackboxsw please take my words with a grain of salt as I'm not a native English speaker.
I think rephrasing the first statement in a positive way reduces the cognitive load a litte bit thus makes it less prone to misunderstandings. The second one is great IMO.

```diff
--- a/doc/rtd/reference/datasources/wsl.rst
+++ b/doc/rtd/reference/datasources/wsl.rst
@@ -51,7 +51,7 @@ User data can be supplied in any
 :ref:`format supported by cloud-init<user_data_formats>`, such as YAML
 cloud-config files or shell scripts. At runtime, the WSL datasource looks for
 user data in the following locations inside the Windows host filesystem, in the
-order specified below.
+order specified below. The WSL datasource will be enabled if cloud-init discovers at least one of the applicable config files described below.
 
 First, configurations from Ubuntu Pro/Landscape are checked for in the
 following paths:
@@ -71,8 +71,8 @@ following paths:
    used instead of the one provided by the ``agent.yaml``, which is treated as
    a default.
 
-Then, if a file from (1) is not found, a user-provided configuration will be
-looked for instead in the following order:
+Then, if a file from (1) is not found, optional user-provided configuration
+will be looked for in the following order:
 
 1. ``%USERPROFILE%\.cloud-init\<InstanceName>.user-data`` holds user data for a
    specific instance configuration. The datasource resolves the name attributed
```

**Comment by blackboxsw** [comment]
> @blackboxsw please take my words with a grain of salt as I'm not a native English speaker. I think rephrasing the first statement in a positive way reduces the cognitive load a litte bit thus makes it less prone to misunderstandings. The second one is great IMO.

LGTM!

**Review by blackboxsw — COMMENTED** [review]
Thx for the update @aciba90. And @CarlosNihelton for policy suggestions and doc suggestions.

I think what remains is unittest coverage of this case and doc additions Carlos mentioned. No need in this case to raise InvalidMetaDataException as we are in `_get_data` and will correctly `return False` below to indicate that the WSL datasource is inapplicable if we have neither agent_data nor user_data.

**Comment by aciba90** [comment]
Thanks, @blackboxsw and @CarlosNihelton, for the reviews. I have updated the PR including documentation changes and unit tests. Please review it.

File: tests/unittests/sources/test_wsl.py [hunk 1]
```diff
@@ -460,6 +461,53 @@ def test_get_data_jinja(self, m_lsb_release, paths, tmpdir):
...
+        """
+        Assert behavior of empty .cloud-config dir with and without agent data
+        """
+        m_lsb_release.return_value = SAMPLE_LINUX_DISTRO
+        data_path = tmpdir.join(".cloud-init", f"{INSTANCE_NAME}.user-data")
+        data_path.dirpath().mkdir()
+
+        if with_agent_data:
+            ubuntu_pro_tmp = tmpdir.join(".ubuntupro", ".cloud-init")
+            os.makedirs(ubuntu_pro_tmp, exist_ok=True)
```
Comment (blackboxsw) [inline tests/unittests/sources/test_wsl.py:478]: Nit, let's move this dir creation outside of the with_agent_data clause to ensure we are coping with presence of an empty directory as the unit test suggests.

File: tests/unittests/sources/test_wsl.py [hunk 2]
```diff
@@ -460,6 +461,53 @@ def test_get_data_jinja(self, m_lsb_release, paths, tmpdir):
...
+
+        ds = wsl.DataSourceWSL(
+            sys_cfg=SAMPLE_CFG,
+            distro=_get_distro("ubuntu"),
+            paths=paths,
+        )
+
+        assert ds.get_data() is with_agent_data
+        ud = ds.get_userdata(True)
+        print(ud)
```
Comment (blackboxsw) [inline tests/unittests/sources/test_wsl.py]: no need to print(ud) from unittests. will make that change locally and force push.

File: tests/unittests/sources/test_wsl.py [hunk 3]
```diff
@@ -460,6 +461,53 @@ def test_get_data_jinja(self, m_lsb_release, paths, tmpdir):
...
+        m_lsb_release.return_value = SAMPLE_LINUX_DISTRO
+        data_path = tmpdir.join(".cloud-init", f"{INSTANCE_NAME}.user-data")
+        data_path.dirpath().mkdir()
+
+        if with_agent_data:
+            ubuntu_pro_tmp = tmpdir.join(".ubuntupro", ".cloud-init")
+            os.makedirs(ubuntu_pro_tmp, exist_ok=True)
+            agent_path = ubuntu_pro_tmp.join("agent.yaml")
+            agent_path.write(
+                "#cloud-config\nlandscape:\n client: {account_name: agent}\n"
```
Comment (blackboxsw) [inline tests/unittests/sources/test_wsl.py]: Let's just use AGENT_SAMPLE var from above so we can assert that value in ds.userdata_raw without duplicating this specific string below.

File: tests/unittests/sources/test_wsl.py [hunk 4]
```diff
@@ -460,6 +461,53 @@ def test_get_data_jinja(self, m_lsb_release, paths, tmpdir):
...
+            x.message
+            for x in caplog.records
+            if x.levelno == expected_log_level and re.match(regex, x.message)
+        ]
+        assert (
+            len(messages) > 0
+        ), "Expected log message matching '{}' with log level '{}'".format(
+            regex, expected_log_level
+        )
+        assert ud is not None
```
Comment (blackboxsw) [inline tests/unittests/sources/test_wsl.py]: This only asserts that get_userdata() returns a MIME message (which passes for both absent and present userdata). Let's actually assert that correct ds.userdata_raw or None was found for each case instead.

**Review by blackboxsw — APPROVED** [review]
Minor test nits which I've pushed as a separate commit. +1 for merge once unittests pass.

**Comment by aciba90** [comment]
Thanks, @blackboxsw, for the test improvements!

**Pull request merged**